### ISSUE TITLE
TargetLowering::LowerAsmOperandForConstraint doesn't handle 2D+ arrays like X86TargetLowering::LowerAsmOperandForConstraint

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#469]
as reported in https://github.com/ClangBuiltLinux/linux/issues/6#issuecomment-486877672 by @nathanchance ,

I see an error in

> ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- make CC=clang -j46 net/bridge/br_forward.o

seems like the use of `static_key_false` in `nf_hook` in `include/linux/netfilter.h` is related, somehow.

**Comment by nathanchance** [comment]
I might be a good idea to keep a list of asm goto patches to apply to LLVM for everything to properly work (whether from Phabricator or whatever).

**Comment by nickdesaulniers** [comment]
```sh
#!/bin/bash
set -ex

# %c support
git cherry-pick 5277b3ff25185b78dfd060e3bdca4d9bcf40b7c5
git cherry-pick b9a00cb504b6ddf4982ad7995d445b8f19f465b6
git cherry-pick ad8f3a1440c74a0b22ebff47cacab57990d08f74
git cherry-pick 8ec304c9fd92442393fcdd98f08fa07ac3ccf5df
git cherry-pick 3271ca01fe03ba50c26304a18c1008c3ae8b5521
git cherry-pick a2077bab40875a72c4dd3af35c59a00a00a9af2c
git cherry-pick 9609ce2f330543fa66af49af3896c1f64d4e4f9f
git cherry-pick 7ab164c4a427b559a7a47fa62ef365862705f950

# inlining
git cherry-pick 1119084b1c5a0db8d40c1a3b8d8d772aa9de9705

# clang asm goto
git cherry-pick 65a1aa609ebd61d3d936612f6111551bb69da1ee
```

**Comment by nickdesaulniers** [comment]
pretty sure this has something to do with arrays w/ rank > 1.  Changing

```c
extern struct static_key nf_hooks_needed[NFPROTO_NUMPROTO][NF_MAX_HOOKS];
```
to a 1d array (instead of 2d) seems to make it work.  @nathanchance I cannot repro any issues with x86 allyesconfig.

**Comment by nathanchance** [comment]
x86 worked fine for me as well, it was every other architecture that blew up.

I will try to apply the latest version of all patches when I get home in a couple of hours and test again (and give full errors rather than the one line that my scripts grep for).

**Comment by nickdesaulniers** [comment]
Ok, I think I understand what's going wrong and how to fix this.

`"i"` in inline asm is handled by `TargetLowering::LowerAsmOperandForConstraint()`, a virtual method that's part of SelectionDAG.  The overrides for arm64 (and probably others) handle machine specific constraints, then defer to the base classes method.  `AArch64TargetLowering::LowerAsmOperandForConstraint` doesn't handle "i", so it defers to the base class.  The base class handles calculating offsets from arrays of structs.

BUT! it does not handle multiple additions (array of array of structs).  Turns out `X86TargetLowering::LowerAsmOperandForConstraint` does:
1. handle "i"
2. correctly calculate offsets of offsets

which is critical since `getelementpointer` can be arbitrarily deep.  That is why we don't see the bug on x86.

The solution is to move the correct `getelementpointer` addition code from `X86TargetLowering::LowerAsmOperandForConstraint` to `TargetLowering::LowerAsmOperandForConstraint()` (kind of like how for `%c` code that should have been arch independent was implemented in an arch dependent way).  I think that will solve the bug.

As to why I thought this was a regression, I think I reported asm goto working for x86, but misremembered testing on arm64 (I likely did not).  When I discovered #449 , I likely thought that there was one bug, when in reality it was two that happened to have the same symptom/produce the same warning.

Will fix next week.
